lua/configs/heirline.lua

```diff
@@ -107,6 +206,54 @@ local heirline_opts = astronvim.user_plugin_opts("plugins.heirline", {
     },
     astronvim.status.component.breadcrumbs { hl = { fg = "winbar_fg", bg = "winbar_bg" } },
   },
+  vim.g.heirline_bufferline -- TODO v3: remove this option and make bufferline default
+      and { -- bufferline
+        { -- file tree padding
+          condition = function(self)
+            self.winid = vim.api.nvim_tabpage_list_wins(0)[1]
+            return vim.tbl_contains({ "neo-tree", "NvimTree" }, vim.bo[vim.api.nvim_win_get_buf(self.winid)].filetype)
+          end,
+          provider = function(self) return string.rep(" ", vim.api.nvim_win_get_width(self.winid)) end,
+          hl = { bg = "bg" },
+        },
+        -- buffer list component
+        astronvim.status.heirline.make_buflist(astronvim.status.component.file_info {
+          file_icon = { padding = { left = 1 } },
+          unique_path = { hl = { fg = "winbarnc_fg" } },
+          close_button = {
+            hl = { fg = "close_fg" },
+            padding = { left = 1, right = 1 },
+            on_click = {
+              callback = function(_, minwid) astronvim.close_buf(minwid) end,
+              minwid = function(self) return self.bufnr end,
+              name = "heirline_tabline_close_buffer_callback",
+            },
+          },
+          padding = { left = 1, right = 1 },
+          hl = function(self)
+            return {
+              fg = self.is_active and "tab_fg" or "tab_inactive_fg",
+              bold = self.is_active,
+              italic = self.is_active,
```

Comment:
Is there a way to control the bold/italic choice? I did see `user_plugin_opts` being called to control these options.

Reply:
Yup! You can customize it just like the statusline using the status API we provide. Here is my own personal customer bufferline that disables the close button as an example: https://code.mehalter.com/AstroNvim_user/~files/master/plugins/heirline.lua

Reply:
Everything can be configured completely from the `plugins.heirline`

Reply:
Thanks, I understand now. Because the config is a parameter to a function am I right in thinking that the entire bit of that config must be replicated except for the bit to be modified? 

So, `astronvim.status.heirline.make_buflist(astronvim.status.component.file_info {...}` the entire table must be populated.

Reply:
Yup!

Reply:
I think you are following me! So to remove the `italic` setting I replicate all of `config[3][2]` and remove the italic line. It would be nice if there was a simpler way of doing that, which admittedly I don't know what that is.

Some of what I'm thinking here is v3 stuff, but let me spit ball a concept. Don't know if it's consistent with the philosophy of the framework, but let me try.

For readability, change here: https://github.com/AstroNvim/AstroNvim/blob/nightly/lua/configs/heirline.lua#L164-L165 to:
```
local heirline_opts = astronvim.user_plugin_opts("plugins.heirline", {
  statusline = {...},
  winbar = {...},
  bufferline = {...},
}
heirline.setup(heirline_opts.statusline, heirline_opts.winbar, heirline_opts.bufferline)
```

In each of the 3 sections above, keep what is in the table unchanged (the ... stuff), except add another table, perhaps called `opts` or `astro_opts`. Then add a whatever "tweaking" options that would be useful, like say `italic` ;)  So for example we end up with:
```
local heirline_opts = astronvim.user_plugin_opts("plugins.heirline", {
  ...
  bufferline = {
    astro_opts = {
      italic = false,
    }
  }
}
```
Anyway, I hope this provokes some thoughts. Much of the other config are tables being passed into functions and single entries in those tables can be modified so the user ends up with a "tweaked" config. In the bufferline case the tweaking is a function that has substantial enough logic that I'm repeating core framework code.

Reply:
@gwww the latest push exposes a bunch of new options to help the user easily fine tune styles ! It isn't done through the `plugins.heirline` table since that strays from the philosophy of `plugins.X = {}` should directly overwrite the `require("X").setup({})` call. We have put it into a new table called `heirline.bufferline` in `user/init.lua` (or `user/heirline/bufferline.lua` file if splitting up config):

All Options:

```lua
return {
  heirline = {
    bufferline = {
      -- true, false, or conditional function to calculate when to color filetype icons
      -- example to color only active and visible buffers:
      -- color_icons = function(self) return self.is_active or self.is_visible end
      color_icons = true,
      -- styling choices for each element, check possible attributes with `:h attr-list`
      buffer = {},
      buffer_path = {},
      buffer_close = {},
      buffer_visible = {},
      buffer_visible_path = {},
      buffer_visible_close = {},
      buffer_active = { bold = true, italic = true },
      buffer_active_path = {},
      buffer_active_close = {},
      buffer_picker = { bold = true },
      buffer_overflow = {},
      tab = {},
      tab_active = {},
      tab_close = {},
    },
  },
}
```

For example, if you wanted to disable the italics for the active buffer, you would simply need this:

```lua
return {
  heirline = {
    bufferline = {
      buffer_active = { italic = false },
    },
  },
}
```

the `color_icons` one is nice as well, it either lets you enable/disable filetype icon colors with a boolean (true/false) or you can do it with a conditional function. So if you wanted to only color the filetype icons for active or visible buffers:

```lua
return {
  heirline = {
    bufferline = {
      color_icons = function(self) return self.is_active or self.is_visible end
    },
  },
}
```

Reply:
Let me know what you think of this @gwww :) Tried to get some easy tweaking added into the setup while still aligning with the philosophy

Reply:
I like the approach, although I have not got it to work yet. I created `user/heirline/bufferline.lua` with the disable italic example above and I'm still getting italics. I'm still playing and will read code further to see what is going on (i.e.: what I messed up!).

For a moment, looking at the bigger picture, it feels like config is split oddly. For example, I now have (assuming I've got the right path):
- `user/plugins/heirline.lua` to tweak my statusline
- `user/heirline/bufferline.lua` to tweak my bufferline

I understand how it got this way, feels like an inconsistent story (philosophy) being told. Does that make sense?

Reply:
If you want to modify the structure of the statusline, winbar, or tabline then you will need to modify `plugins.heirline`. 

If you want to modify exposed options they will go into `heirline.X` we currently have: `heirline.colors`, `heirline.separators`, and `heirline.bufferline`.

You said you haven't figured out how to make it work yet, what's currently in your file for changing it?

Reply:
@gwww if you are putting it into a separate file (`user/heirline/bufferline.lua`) then it should look like this:

```lua
return {
  buffer_active = { italic = false },
}
```

Reply:
Maybe we should change it to be more general and add styles to the statusline elements: thinking 🤔 then we have colors and styles to go with each color

Reply:
I was also thinking about the `highlight` functionality, then **styling** *[line|bar] is just like other stuff. There will still be other options, such as `color_icons` that will have to fit into the config.

BTW, I did get mine working... I haven't had much time to look at anything today. I had a typo.

Reply:
@gwww just a heads up, I revised the previous commit and moved it to `heirline.attributes` so you will need to rename your `heirline/bufferline.lua` file to `heirline/attributes.lua`. I still need to rework the options like `color_icons` to fit nicely into the philosophy, but this now works for all element names :)

Reply:
@gwww ok, I just finished cleaning this up. Here is an example of the current `heirline` table options

```lua
  -- Customize Heirline options
  heirline = {
    -- Customize different separators between sections
    separators = {
      tab = { "", "" },
    },
    -- Customize colors for each element each element has a `_fg` and a `_bg`
    colors = function(colors)
      colors.git_branch_fg = astronvim.get_hlgroup "Conditional"
      return colors
    end,
    -- Customize attributes of highlighting in Heirline components
    attributes = {
      -- styling choices for each heirline element, check possible attributes with `:h attr-list`
      git_branch = { bold = true }, -- bold the git branch statusline component
      buffer_active = { italic = false }, -- disable italic of active buffer
    },
    -- Customize if icons should be highlighted
    icon_highlight = {
      -- disable highlighting LSP symbols in the breadcrumbs
      breadcrumbs = false,
      file_icon = {
        -- highlight filetype icon in the statusline
        statusline = true,
        -- Only color tabline filetype icons if the buffer is active or visible
        tabline = function(self) return self.is_active or self.is_visible end, 
      },
    },
  },
```

Reply:
@gwww just a heads up I renamed the `icon_highlight` table to `icon_highlights` to match the rest of the plurality in the names

Reply:
Thx